### ISSUE TITLE
Truthiness of object arrays

### ISSUE THREAD

**Issue by anntzer** [numpy/numpy#5276]
Initially discussed on numpy-discussion.

I am puzzled by the following (numpy 1.9.0, python 3.4.2):

```
In [1]: t = array(None); t[()] = array([None, None])  # Construct a 0d array of dtype object, containing a single numpy array with 2 elements

In [2]: bool(t), t.__bool__()
Out[2]: (True, True)

In [3]: if t: pass
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

I thought that `if x` simply calls `bool`, but apparently this is not even the case...

**Comment by anntzer** [comment]
On 3.5, calling `bool` itself directly fails with
```
In [2]: t = array(None); t[()] = array([None, None]); bool(t)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()

During handling of the above exception, another exception occurred:

SystemError                               Traceback (most recent call last)
<ipython-input-2-1cd478a6057e> in <module>()
----> 1 t = array(None); t[()] = array([None, None]); bool(t)

SystemError: <class 'bool'> returned a result with an error set
```
which I guess is a progress on Python's side.

**Comment by eric-wieser** [comment]
I think this was fixed in #9300?
